data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+use serde::{Serialize, Serializer};
+
+const MAX_INLINED_STRING_LEN: usize = 2 * size_of::<usize>() - 1;
+
+/// Immutable inlinable string.
+/// Strings shorter than 15/7/3 bytes (depending on architecture 64/32/16 bit pointer width) are inlined.
+#[derive(Clone)]
+pub struct ConstString(ConstStringData);
+
+impl ConstString {
+    /// Returns the length of this [`ConstString`], in bytes.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if [`ConstString`] is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.len() == 0
+    }
+
+    /// Returns `true` if [`ConstString`] data is inlined.
+    #[inline]
+    pub const fn is_inlined(&self) -> bool {
+        self.0.is_inlined()
+    }
+
+    /// Construct empty [`ConstString`].
+    #[inline]
+    pub const fn new() -> Self {
+        Self(ConstStringData::new())
+    }
+}
+
+impl Deref for ConstString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl<T: ?Sized> AsRef<T> for ConstString
+where
+    ConstStringData: AsRef<T>,
+{
+    #[inline]
+    fn as_ref(&self) -> &T {
+        self.0.as_ref()
+    }
+}
+
+impl<T: Into<ConstStringData>> From<T> for ConstString {
+    #[inline]
+    fn from(value: T) -> Self {
+        Self(value.into())
+    }
+}
+
+impl Display for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{}", &**self)
+    }
+}
+
+impl Debug for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{:?}", &**self)
+    }
+}
+
+impl Hash for ConstString {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (**self).hash(state)
+    }
+}
+
+impl PartialOrd for ConstString {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        PartialOrd::partial_cmp(&**self, &**other)
+    }
+}
+
+impl Ord for ConstString {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ord::cmp(&**self, &**other)
+    }
+}
+
+impl PartialEq for ConstString {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        PartialEq::eq(&**self, &**other)
+    }
+}
+
+macro_rules! impl_eq {
+    ($($ty:ty,)*) => {
+        impl_eq!($($ty),*);
+    };
+    ($($ty:ty),*) => {$(
+        impl PartialEq<$ty> for ConstString {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &$ty) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+
+        impl PartialEq<ConstString> for $ty {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &ConstString) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+    )*};
+}
+
+impl_eq!(String, str, &str);
+
+impl Eq for ConstString {}
```

Comment:
derive it

Reply:
rust requires all fields to derive `Eq` trait in that case